build: support RHEL/CentOS 6

tamird opened the pull request:
- cockroachdb/builder is now based on ubuntu xenial
- cockroachdb/builder targets RHEL/CentOS 6 versions of linux and glibc
- cockroachdb/builder is upgraded to Go 1.7.5
- bootstrap-debian is mostly absorbed into cockroachdb/builder
- STATIC=1 is replaced with TYPE=release
- a new Makefile option (type=static) produces fully static binaries by
  linking to musl instead of glibc
- `cockroach version` now reports the build type (release/static)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13102)
<!-- Reviewable:end -->

petermattis commented:
Nice that you got this working, though I'm mildly anxious about merging this. We have a significant body of C++ code that uses libc/libpthread that I worry this will introduce instability. I'm open to hearing the other side if you think this concern is misplaced.

I think we should run a musl-based build on one or more of the test clusters for a decent amount of time. Can you pull out the non-contentious parts of this change into a separate PR?

a-robinson commented:
So would this effectively switch all our release builds to use musl?

Like @petermattis, I'm very wary of switching over to musl for any sort of wide usage. Being able to support it is great, but how many other widely used projects default to it? We'll at the very least get a handful of questions about DNS resolver behavior, and likely more issues that'll take up our time. Perhaps we should talk about this tomorrow.

bdarnell commented:
I'm also wary of switching to musl all at once, although if we try to offer the two builds in parallel will we get a meaningful amount of testing of the musl build?

At a minimum a change like this should be sequenced after the rocksdb 5.0/DeleteRange upgrade. 

> We'll at the very least get a handful of questions about DNS resolver behavior

I think that if we get even a handful of questions about DNS then we should stick with a glibc build as our primary offering and offer a musl build only as an alternative for rhel/centos/alpine users. The theory behind switching is that musl has caught up with glibc's resolver in the ways that matter (the specific issue that caused problems for us last time we tried this, gliderlabs/docker-alpine#8, has been fixed)  and deployments that need other glibc-specific features are rare enough that we can ask those users to build from source (note that this is speculation with no real data behind it. I'm much more concerned about RHEL users than users of alpine or other musl-based distros, so I'd be most comfortable with an outcome in which we ship builds for an old version of glibc instead of static musl-based builds. However, getting the combination of an old version of glibc with new versions of compilers and other dependencies appears to require quite a lot of yak shaving). 

---

Reviewed 4 of 4 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 13 of 13 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Makefile, line 67 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbDaqzypSf3PBC3_Pzx:-KbDaqzzgumRFPlmT0E2:bhmqpkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/45601b0d319e916f4457ae83628575c11c840815/Makefile#L67)):*
> ```Makefile
> # and link against it instead.
> TAGS += musl
> LDFLAGS += -extldflags '-static'
> ```

Those quotes don't do anything, do they?

---

*[build/Dockerfile, line 20 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbDb_u-cyodhPbjYz1l:-KbDb_u-cyodhPbjYz1m:b-y5crsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/45601b0d319e916f4457ae83628575c11c840815/build/Dockerfile#L20)):*
> ```Dockerfile
>     bzip2 \
>     curl \
>     fontconfig \
> ```

While you're adding comments, why do we need `fontconfig`?

---

*[scripts/gceworker.sh, line 25 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbDbCBw9zNEhT4ouxSj:-KbDbCBw9zNEhT4ouxSk:bkdhiqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/45601b0d319e916f4457ae83628575c11c840815/scripts/gceworker.sh#L25)):*
> ```Shell
>     sleep 20 # avoid SSH timeout on copy-files
> 
>     gcloud compute copy-files "$(dirname "${0}")" "${name}:scripts"
> ```

Both `gceworker` and `azworker` need to be updated to copy the parallel-build patches from their new locations. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-:-KbDdhbm-X-JYqJEhPmP:b-5o2vh)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Makefile, line 67 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbDaqzypSf3PBC3_Pzx:-KbEjWOXzbyxvrAqTQ-n:b-v2wy36) ([raw file](https://github.com/cockroachdb/cockroach/blob/45601b0d319e916f4457ae83628575c11c840815/Makefile#L67)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Those quotes don't do anything, do they?
</blockquote></details>

Probably not; removed.

---

*[build/Dockerfile, line 20 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbDb_u-cyodhPbjYz1l:-KbEjZRMiH-CFOGo04Ul:b4roqke) ([raw file](https://github.com/cockroachdb/cockroach/blob/45601b0d319e916f4457ae83628575c11c840815/build/Dockerfile#L20)):*
There's already a comment covering that

> bzip2 and fontconfig are used by phantomjs-prebuilt to test the UI.

---

*[scripts/gceworker.sh, line 25 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbDbCBw9zNEhT4ouxSj:-KbEki7OtwouHdgy2vh_:b2tghrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/45601b0d319e916f4457ae83628575c11c840815/scripts/gceworker.sh#L25)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Both `gceworker` and `azworker` need to be updated to copy the parallel-build patches from their new locations. 
</blockquote></details>

Instead, could we just bake everything we need into the builder image? Unlike before, the builder now has a go toolchain which includes the parallel-build patches.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 10 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[build/Dockerfile, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbFvqRLaELuepuFQ8jl:-KbFvqRMulwiICehrz6k:b4bxwzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc52e3858b1e2599c316033823146232fb7c3/build/Dockerfile#L1)):*
> ```Dockerfile
> FROM alpine:3.5
> ```

Perhaps this should be a `Dockerfile-musl` along with a separate builder image. Seems like a few more additions to this PR and you could make the `musl` builds an option so we could get this PR merged without changing the existing static builds.

---

*[scripts/gceworker.sh, line 25 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbDbCBw9zNEhT4ouxSj:-KbFvTVJG8dn_uYx42B0:b-8fbfie) ([raw file](https://github.com/cockroachdb/cockroach/blob/45601b0d319e916f4457ae83628575c11c840815/scripts/gceworker.sh#L25)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Instead, could we just bake everything we need into the builder image? Unlike before, the builder now has a go toolchain which includes the parallel-build patches.
</blockquote></details>

And use the builder image on the `{gce,az}worker`? I don't think we have docker set up on the worker images, and using docker in that fashion feels like it would be a pain.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 10 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[build/Dockerfile, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbFvqRLaELuepuFQ8jl:-KbFwYJs5KHR8odUaK3t:bdzt90a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc52e3858b1e2599c316033823146232fb7c3/build/Dockerfile#L1)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Perhaps this should be a `Dockerfile-musl` along with a separate builder image. Seems like a few more additions to this PR and you could make the `musl` builds an option so we could get this PR merged without changing the existing static builds.
</blockquote></details>

Yeah, I can do that. I'd rather not, but I will.

---

*[scripts/gceworker.sh, line 25 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbDbCBw9zNEhT4ouxSj:-KbFwRangqlCInzanA_S:b6o1f64) ([raw file](https://github.com/cockroachdb/cockroach/blob/45601b0d319e916f4457ae83628575c11c840815/scripts/gceworker.sh#L25)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

And use the builder image on the `{gce,az}worker`? I don't think we have docker set up on the worker images, and using docker in that fashion feels like it would be a pain.
</blockquote></details>

We definitely have docker on the workers; see bootstrap-debian. I used that workflow heavily in preparing this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 10 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[build/Dockerfile, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbFvqRLaELuepuFQ8jl:-KbFxekdqgYGy_uXuKDF:bbhlbsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc52e3858b1e2599c316033823146232fb7c3/build/Dockerfile#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Yeah, I can do that. I'd rather not, but I will.
</blockquote></details>

Well, the alternative is to hold off on merging this PR for a while. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

tamird committed:
build: support RHEL/CentOS 6

- cockroachdb/builder is now based on ubuntu xenial
- cockroachdb/builder targets RHEL/CentOS 6 versions of linux and glibc
- cockroachdb/builder is upgraded to Go 1.7.5
- bootstrap-debian is mostly absorbed into cockroachdb/builder
- STATIC=1 is replaced with TYPE=release
- a new Makefile option (type=static) produces fully static binaries by
  linking to musl instead of glibc
- `cockroach version` now reports the build type (release/static)

tamird commented:
This has run into an unexpected snag; because TC workers have some cached build artifacts that have been linked against debian:jessie's glibc, building CR with this new image causes missing symbol errors when those artifacts are linked with ubuntu:xenial's glibc. Normally, the solution to this would be to use a new installsuffix, but our TC workers are also currently suffering from disk space issues which would be exacerbated by the addition of a new installsuffix.

Also, we still need a reasonable way of toggling between musl and glibc distributions.

tamird commented:
This PR needs a new title and description, but is otherwise ready to be re-reviewed. The new approach produces binaries dynamically linked to glibc (but otherwise static), with a separate `make` option to produce fully static binaries with musl.

@bdarnell @petermattis @mberhault

bdarnell commented:
Reviewed 2 of 13 files at r6, 2 of 15 files at r8, 2 of 9 files at r9, 19 of 19 files at r10.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[Makefile, line 23 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvEUdcPft1wS-VXS0_:-KbvEUdcPft1wS-VXS0a:bjn066m) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/Makefile#L23)):*
> ```Makefile
> # Allow setting of go build flags from the command line.
> GOFLAGS :=
> # Possible values:
> ```

Comment that both RELEASE and STATIC builds only work within the builder image. Comments about why these particular versions were chosen (even if it's just "these are the oldest versions available with the tools we're using") would probably be helpful the next time we change any of this.

Nit: I slightly prefer lowercase names for the values of TYPE.

---

*[Makefile, line 66 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvFW9MyxcKOo9MtyJN:-KbvFW9NFo9JrJtlvlQ_:b-d570lc) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/Makefile#L66)):*
> ```Makefile
> export CC  = /x-tools/x86_64-unknown-linux-gnu/bin/x86_64-unknown-linux-gnu-gcc
> export CXX = /x-tools/x86_64-unknown-linux-gnu/bin/x86_64-unknown-linux-gnu-g++
> override LDFLAGS += -extldflags "-static-libgcc -static-libstdc++"
> ```

We should probably also set a `-X github.com/cockroachdb/cockroach/pkg/build.type=release` flag so that the build type can show up in `cockroach version`. 

---

*[Makefile, line 67 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvHt6QZCwGpHZ7qb38:-KbvHt6QZCwGpHZ7qb39:bh2ytym) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/Makefile#L67)):*
> ```Makefile
> export CXX = /x-tools/x86_64-unknown-linux-gnu/bin/x86_64-unknown-linux-gnu-g++
> override LDFLAGS += -extldflags "-static-libgcc -static-libstdc++"
> override GOFLAGS += -installsuffix glibc
> ```

Nit: `release` instead of `glibc`?

---

*[build/builder.sh, line 115 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvHp4Uc6OjrA2_9ek4:-KbvHp4VmqIURXlL1o2U:b-4q8cwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/builder.sh#L115)):*
> ```Shell
> vols="${vols} --volume=${gopath0}/pkg/docker_amd64:/go/pkg/linux_amd64"
> vols="${vols} --volume=${gopath0}/pkg/docker_amd64_glibc:/go/pkg/linux_amd64_glibc"
> vols="${vols} --volume=${gopath0}/pkg/docker_amd64_gmusl:/go/pkg/linux_amd64_musl"
> ```

Is `gmusl` a typo?

---

*[build/Dockerfile, line 78 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvJLczlNfUMaB8gY3O:-KbvJLd-5hdCIqTAA1rD:bb77rlh) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L78)):*
> ```Dockerfile
>  && gpg --list-keys \
>  && gpg --recv-keys 35B871D1 \
>  && gpg --verify crosstool-ng-1.22.0.tar.bz2.sig \
> ```

Yay!

---

*[build/Dockerfile, line 98 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvLT2O1qiHfg537xI0:-KbvLT2PUXKrFIGiPBDM:bt6rmno) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L98)):*
> ```Dockerfile
> RUN groupadd -r roachbuild && useradd -m -r -g roachbuild roachbuild
> RUN mkdir /x-tools && chown roachbuild:roachbuild /x-tools
> USER roachbuild
> ```

Why make a new user just for this part?

---

*[build/Dockerfile, line 110 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvKwn3-IMFJRRhluJO:-KbvKwn3-IMFJRRhluJP:bghxr81) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L110)):*
> ```Dockerfile
> USER root
> 
> RUN apt-get autoremove -y g++
> ```

Why do we need this?

---

*[build/Dockerfile, line 118 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvL4Qzmlfkndqafo7i:-KbvL4Qzmlfkndqafo7j:be1yltr) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L118)):*
> ```Dockerfile
>  && mkdir llvm/projects/libcxx    && curl -sfSL http://releases.llvm.org/3.9.1/libcxx-3.9.1.src.tar.xz    | tar --strip-components=1 -C llvm/projects/libcxx -xJ \
>  && mkdir llvm/projects/libcxxabi && curl -sfSL http://releases.llvm.org/3.9.1/libcxxabi-3.9.1.src.tar.xz | tar --strip-components=1 -C llvm/projects/libcxxabi -xJ \
>  && mkdir libcxx_msan && (cd libcxx_msan && cmake ../llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_USE_SANITIZER=Memory && make cxx -j16)
> ```

`-j$(nproc)`

---

*[build/Dockerfile, line 123 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvL9tGPLsZcu7QnsAF:-KbvL9tGPLsZcu7QnsAG:b2kwv0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L123)):*
> ```Dockerfile
> 
> COPY parallelbuilds-go1.7.patch /
> RUN curl -fsSL https://golang.org/dl/go1.7.5.src.tar.gz -o golang.tar.gz \
> ```

Mention the upgrade to Go 1.7.5 in the commit message.

---

*[scripts/azworker.sh, line 14 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMEdpFQ8rhLK3Yila:-KbvMEdpFQ8rhLK3Yilb:bubq6z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/scripts/azworker.sh#L14)):*
> ```Shell
> USER="${USER-$(id -un)}"
> CLUSTER="azworker-${USER}"
> NAME=${AZWORKER_NAME-${CLUSTER}-azworker}
> ```

Remove the `-azworker`; it's already included in `${CLUSTER}`

---

*[scripts/with-msan, line 9 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMcoib1b56Jt4fD_9:-KbvMcoib1b56Jt4fD_A:b-1egzw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/scripts/with-msan#L9)):*
> ```
> #
> # MemorySanitizer is only available on linux; this script assumes the machine has
> # been set up with bootstrap-debian.sh.
> ```

Change this comment to say that it assumes it's running in the builder container.

---

*[scripts/with-msan, line 12 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMhrfEqgGMKFO32b_:-KbvMhrfEqgGMKFO32ba:b-2y309o) ([raw file](https://github.com/cockroachdb/cockroach/blob/621dfa6c27b1e64be2a0d88eb54dcf7c5267742a/scripts/with-msan#L12)):*
> ```
> 
> export CC=clang
> export CXX=clang++
> ```

With these lines gone, isn't it using gcc instead of clang?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[Makefile, line 23 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvEUdcPft1wS-VXS0_:-KbvUmHlD-aozcWzD-I8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/Makefile#L23)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Comment that both RELEASE and STATIC builds only work within the builder image. Comments about why these particular versions were chosen (even if it's just "these are the oldest versions available with the tools we're using") would probably be helpful the next time we change any of this.

Nit: I slightly prefer lowercase names for the values of TYPE.
</blockquote></details>

Done.

---

*[Makefile, line 66 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvFW9MyxcKOo9MtyJN:-KbvUn1RAAuApmesysMU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/Makefile#L66)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We should probably also set a `-X github.com/cockroachdb/cockroach/pkg/build.type=release` flag so that the build type can show up in `cockroach version`. 
</blockquote></details>

Done.

---

*[Makefile, line 67 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvHt6QZCwGpHZ7qb38:-KbvUnZdg28ULZzO0RhX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/Makefile#L67)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Nit: `release` instead of `glibc`?
</blockquote></details>

Done.

---

*[build/builder.sh, line 115 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvHp4Uc6OjrA2_9ek4:-KbvV2ejQfqPQCEbR9Zc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/builder.sh#L115)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is `gmusl` a typo?
</blockquote></details>

Done.

---

*[build/Dockerfile, line 98 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvLT2O1qiHfg537xI0:-KbvPSFX-uXN9_IIHyNp:bww73q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L98)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why make a new user just for this part?
</blockquote></details>

https://github.com/crosstool-ng/crosstool-ng/commit/1cabb74b7d629262319c9e348962088a0697fe33

Added a comment.

---

*[build/Dockerfile, line 110 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvKwn3-IMFJRRhluJO:-KbvPWZArZXAhvHbHpCw:b-vnvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L110)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why do we need this?
</blockquote></details>

To avoid having two toolchains (clang is also present). As mentioned in the mega comment above, g++ is for bootstrapping the cross compilation toolchains.

---

*[build/Dockerfile, line 118 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvL4Qzmlfkndqafo7i:-KbvPMM4FkqdPWg5c7nk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L118)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`-j$(nproc)`
</blockquote></details>

Done.

---

*[build/Dockerfile, line 123 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvL9tGPLsZcu7QnsAF:-KbvQXrudsNNb339xVXz:bcquwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L123)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Mention the upgrade to Go 1.7.5 in the commit message.
</blockquote></details>

Done (will squash it into the resulting commit).

---

*[scripts/azworker.sh, line 14 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMEdpFQ8rhLK3Yila:-KbvRFkbCXxOWSVnkxJW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/scripts/azworker.sh#L14)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Remove the `-azworker`; it's already included in `${CLUSTER}`
</blockquote></details>

Done.

---

*[scripts/with-msan, line 9 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMcoib1b56Jt4fD_9:-KbvRRUn0E0DpXvrBFbh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/scripts/with-msan#L9)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Change this comment to say that it assumes it's running in the builder container.
</blockquote></details>

Done.

---

*[scripts/with-msan, line 12 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMhrfEqgGMKFO32b_:-KbvRRyfKBYemLY31Iho:bk0zys6) ([raw file](https://github.com/cockroachdb/cockroach/blob/621dfa6c27b1e64be2a0d88eb54dcf7c5267742a/scripts/with-msan#L12)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

With these lines gone, isn't it using gcc instead of clang?
</blockquote></details>

gcc is no longer available in the build container, so clang is used instead.

I'd consider going the other way - restoring these lines and defaulting the build container to target glibc 2.12, but the cross toolchain is terribly slow for some reason.

Defaulting to the glibc 2.12 toolchain also has the benefit of producing correct `./configure` results when updating `c-*` projects. Perhaps we should change to that until slow compilation becomes a problem. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 13 files at r6, 1 of 9 files at r9, 2 of 19 files at r10, 1 of 19 files at r11, 2 of 7 files at r12, 9 of 17 files at r13, 15 of 15 files at r14.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[build/Dockerfile, line 110 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvKwn3-IMFJRRhluJO:-Kbvc9OOzjf1_2eX820F:bsspmdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L110)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

To avoid having two toolchains (clang is also present). As mentioned in the mega comment above, g++ is for bootstrapping the cross compilation toolchains.
</blockquote></details>

Oh, I see. So release/musl builds differ from the default configuration in which compiler they use, not just which version of libc they target. I'm wary of switching the default from gcc to clang like this. 

---

*[scripts/with-msan, line 12 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMhrfEqgGMKFO32b_:-KbvgfhU1lMu4b6xxPdN:batt7kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/621dfa6c27b1e64be2a0d88eb54dcf7c5267742a/scripts/with-msan#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

gcc is no longer available in the build container, so clang is used instead.

I'd consider going the other way - restoring these lines and defaulting the build container to target glibc 2.12, but the cross toolchain is terribly slow for some reason.

Defaulting to the glibc 2.12 toolchain also has the benefit of producing correct `./configure` results when updating `c-*` projects. Perhaps we should change to that until slow compilation becomes a problem. What do you think?
</blockquote></details>

Ah. Probably still good to be explicit about the compiler, though.

How slow is "terribly slow"? We go out of our way to speed up the builds for `{gce,az}workers` with the parallel cgo patch, so we don't want to slow things down too much. If the crosstool gcc is significantly slower than the one we get from `apt`, then I'd rather have that available too for default builds (even though it is a fourth toolchain and a third build of gcc). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[build/Dockerfile, line 110 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvKwn3-IMFJRRhluJO:-KbvoMv0haUDxWCkXwiu:buo5fjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L110)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Oh, I see. So release/musl builds differ from the default configuration in which compiler they use, not just which version of libc they target. I'm wary of switching the default from gcc to clang like this. 
</blockquote></details>

What's the concern? We're already shipping OSX binaries with clang.

---

*[scripts/with-msan, line 12 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMhrfEqgGMKFO32b_:-KbvnFY6fUMbbLT9WcwJ:b-65fge) ([raw file](https://github.com/cockroachdb/cockroach/blob/621dfa6c27b1e64be2a0d88eb54dcf7c5267742a/scripts/with-msan#L12)):*
> Ah. Probably still good to be explicit about the compiler, though.

Sure, done.

> How slow is "terribly slow"? We go out of our way to speed up the builds for {gce,az}workers with the parallel cgo patch, so we don't want to slow things down too much. If the crosstool gcc is significantly slower than the one we get from apt, then I'd rather have that available too for default builds (even though it is a fourth toolchain and a third build of gcc).

Well, it is available (though it's currently clang, not gcc), but the question is should we use it by default in the build container; doing so requires special handling for default builds; not doing so requires special handling for dependency upgrades.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 23 of 24 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[build/Dockerfile, line 110 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvKwn3-IMFJRRhluJO:-KbvoxPqD8yosBX2jL29:b4yyclv) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L110)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What's the concern? We're already shipping OSX binaries with clang.
</blockquote></details>

Just fear of change, and of differences between regular and release builds. The fact that jemalloc has options referring to "libgcc" and "gcc intrinsics" give me pause (both of these options at least *compile* with clang, but do they work? I don't think we've tried heap profiling on a clang/linux build). 

---

*[scripts/with-msan, line 12 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMhrfEqgGMKFO32b_:-Kbvqn3__IaCSYfsY-Tv:b-up5t14) ([raw file](https://github.com/cockroachdb/cockroach/blob/621dfa6c27b1e64be2a0d88eb54dcf7c5267742a/scripts/with-msan#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> Ah. Probably still good to be explicit about the compiler, though.

Sure, done.

> How slow is "terribly slow"? We go out of our way to speed up the builds for {gce,az}workers with the parallel cgo patch, so we don't want to slow things down too much. If the crosstool gcc is significantly slower than the one we get from apt, then I'd rather have that available too for default builds (even though it is a fourth toolchain and a third build of gcc).

Well, it is available (though it's currently clang, not gcc), but the question is should we use it by default in the build container; doing so requires special handling for default builds; not doing so requires special handling for dependency upgrades.
</blockquote></details>

What do you mean by special handling here? I'd prefer to minimize the difference between the default and release builds (ideally, the only difference would be the use of static vs dynamic linking)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 20 of 25 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[build/x86-linux-gnu.config, line 385 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbwgOUIFLmGqzQ-Kkkr:-KbwgOUIFLmGqzQ-Kkks:b-305yki) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb133eab32ed99d68f3ab7758c426a7524c4d54f/build/x86-linux-gnu.config#L385)):*
> ```
> CT_CC_GCC_HAS_LIBQUADMATH=y
> CT_CC_GCC_HAS_LIBSANITIZER=y
> CT_CC_GCC_VERSION="4.9.3"
> ```

Wait, so this is using gcc 4.9.3? How does that work? Our contributing.md file says that we don't support gcc (more specifically, g++) prior to 6.0. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 20 of 25 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[build/Dockerfile, line 110 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvKwn3-IMFJRRhluJO:-KbzSF8rUVfzZPu6SX7a:b9peark) ([raw file](https://github.com/cockroachdb/cockroach/blob/44d08b7c622cb2e91538dc93e1b5a0ecf32e42a6/build/Dockerfile#L110)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Just fear of change, and of differences between regular and release builds. The fact that jemalloc has options referring to "libgcc" and "gcc intrinsics" give me pause (both of these options at least *compile* with clang, but do they work? I don't think we've tried heap profiling on a clang/linux build). 
</blockquote></details>

OK; I think we agree that clang should be for msan only, and we should use the release toolchain in all other cases.

---

*[build/x86-linux-gnu.config, line 385 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbwgOUIFLmGqzQ-Kkkr:-KbzRs-V6H-MKv9AWoZA:b-qhhf1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb133eab32ed99d68f3ab7758c426a7524c4d54f/build/x86-linux-gnu.config#L385)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Wait, so this is using gcc 4.9.3? How does that work? Our contributing.md file says that we don't support gcc (more specifically, g++) prior to 6.0. 
</blockquote></details>

4.9.3 is "known good" - see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=48891

---

*[scripts/with-msan, line 12 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbvMhrfEqgGMKFO32b_:-KbzS2tRKTD9BjK3n7O6:bv421cf) ([raw file](https://github.com/cockroachdb/cockroach/blob/621dfa6c27b1e64be2a0d88eb54dcf7c5267742a/scripts/with-msan#L12)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What do you mean by special handling here? I'd prefer to minimize the difference between the default and release builds (ideally, the only difference would be the use of static vs dynamic linking)
</blockquote></details>

Yeah, I agree. The only problem is that the cross toolchain is terrible (empirically) slow. Perhaps I'm building an unoptimized gcc? Ideas welcome, but I'll dig more this afternoon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

tamird commented:
FYI the builder image is now `1.864 GB` :(

tamird committed:
build/builder: allow arbitrary mounting of go caches

Set `GOCACHEPATH` to mount somewhere other than GOPATH.

tamird commented:
Still looking for an LGTM here. @bdarnell?

tamird commented:
Review status: 0 of 28 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[build/Dockerfile, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-KbFvqRLaELuepuFQ8jl:-Kc0qIxa41f_qPJj-iOm:bccaoai) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc52e3858b1e2599c316033823146232fb7c3/build/Dockerfile#L1)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Well, the alternative is to hold off on merging this PR for a while. 
</blockquote></details>

By the way, this is now implemented.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 13 files at r6, 1 of 9 files at r9, 2 of 19 files at r10, 1 of 7 files at r12, 2 of 17 files at r13, 5 of 5 files at r15, 1 of 1 files at r16, 3 of 3 files at r17, 1 of 1 files at r18, 31 of 31 files at r20.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13102#-:-Kc1FIDbmkD_9FRWWSjb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
